File: cpp/src/chips/light/APDS9960.cpp
#include "APDS9960.h"

#ifdef __linux__
#include <unistd.h>
#define DELAY_MS(ms) usleep((ms) * 1000)
#else
#include <Arduino.h>
#define DELAY_MS(ms) delay(ms)
#endif

#ifdef CONFIG_ZEPHYR
#include <zephyr/kernel.h>
#undef DELAY_MS
#define DELAY_MS(ms) k_sleep(K_MSEC(ms))
#endif

APDS9960Minimal::APDS9960Minimal(Transport& transport)
    : _transport(transport) {
    DELAY_MS(6);
    uint8_t id = _read_reg(REG_ID);
    (void)id;
    _write_reg(REG_ENABLE, 0x00);
    _write_reg(REG_ATIME, ATIME_DEFAULT);
    _write_reg(REG_CONTROL, CONTROL_DEFAULT);
    _write_reg(REG_CONFIG2, CONFIG2_DEFAULT);
    _write_reg(REG_ENABLE, 0x03);
    DELAY_MS(210);
}

void APDS9960Minimal::_write_reg(uint8_t reg, uint8_t value) {
    uint8_t buf[2] = { reg, value };
    _transport.write(buf, 2);
}

uint8_t APDS9960Minimal::_read_reg(uint8_t reg) {
    uint8_t val;
    _transport.write_read(&reg, 1, &val, 1);
    return val;
}

uint16_t APDS9960Minimal::_read_reg16_le(uint8_t reg) {
    uint8_t buf[2];
    _transport.write_read(&reg, 1, buf, 2);
    return (uint16_t)buf[0] | ((uint16_t)buf[1] << 8);
}

uint16_t APDS9960Minimal::color_clear() {
    return _read_reg16_le(REG_CDATAL);
}

uint16_t APDS9960Minimal::color_red() {
    return _read_reg16_le(REG_RDATAL);
}

uint16_t APDS9960Minimal::color_green() {
    return _read_reg16_le(REG_GDATAL);
}

uint16_t APDS9960Minimal::color_blue() {
    return _read_reg16_le(REG_BDATAL);
}

void APDS9960Minimal::color(uint16_t& clear, uint16_t& red, uint16_t& green, uint16_t& blue) {
    uint8_t reg = REG_CDATAL;
    uint8_t buf[8];
    _transport.write_read(&reg, 1, buf, 8);
    clear = (uint16_t)buf[0] | ((uint16_t)buf[1] << 8);
    red   = (uint16_t)buf[2] | ((uint16_t)buf[3] << 8);
    green = (uint16_t)buf[4] | ((uint16_t)buf[5] << 8);
    blue  = (uint16_t)buf[6] | ((uint16_t)buf[7] << 8);
}

// APDS9960Full

APDS9960Full::APDS9960Full(Transport& transport)
    : APDS9960Minimal(transport) {}

void APDS9960Full::enable_proximity(bool enabled) {
    uint8_t val = _read_reg(REG_ENABLE);
    if (enabled) val |= 0x04;
    else         val &= ~0x04;
    _write_reg(REG_ENABLE, val);
}

uint8_t APDS9960Full::proximity() {
    return _read_reg(REG_PDATA);
}

void APDS9960Full::enable_wait(bool enabled) {
    uint8_t val = _read_reg(REG_ENABLE);
    if (enabled) val |= 0x08;
    else         val &= ~0x08;
    _write_reg(REG_ENABLE, val);
}

void APDS9960Full::configure_wait(uint8_t wtime, bool wlong) {
    _write_reg(REG_WTIME, wtime);
    uint8_t c1 = _read_reg(REG_CONFIG1);
    if (wlong) c1 |= 0x02;
    else       c1 &= ~0x02;
    c1 = (c1 & 0x03) | 0x60;
    _write_reg(REG_CONFIG1, c1);
}

void APDS9960Full::configure_als(uint8_t atime, uint8_t again) {
    _write_reg(REG_ATIME, atime);
    uint8_t ctrl = _read_reg(REG_CONTROL);
    ctrl = (ctrl & 0xFC) | (again & 0x03);
    _write_reg(REG_CONTROL, ctrl);
}

void APDS9960Full::configure_proximity_led(uint8_t ldrive, uint8_t pgain, uint8_t ppulse, uint8_t pplen) {
    uint8_t ctrl = _read_reg(REG_CONTROL);
    ctrl = ((ldrive & 0x03) << 6) | ((pgain & 0x03) << 2) | (ctrl & 0x03);
    _write_reg(REG_CONTROL, ctrl);
    _write_reg(REG_PPULSE, ((pplen & 0x03) << 6) | (ppulse & 0x3F));
}

void APDS9960Full::set_led_boost(uint8_t boost) {
    uint8_t c2 = _read_reg(REG_CONFIG2);
    c2 = (c2 & 0xCF) | ((boost & 0x03) << 4) | 0x01;
    _write_reg(REG_CONFIG2, c2);
}

void APDS9960Full::als_threshold(uint16_t low, uint16_t high) {
    _write_reg(REG_AILTL, low & 0xFF);
    _write_reg(REG_AILTH, (low >> 8) & 0xFF);
    _write_reg(REG_AIHTL, high & 0xFF);
    _write_reg(REG_AIHTH, (high >> 8) & 0xFF);
}

void APDS9960Full::proximity_threshold(uint8_t low, uint8_t high) {
    _write_reg(REG_PILT, low);
    _write_reg(REG_PIHT, high);
}

void APDS9960Full::set_persistence(uint8_t ppers, uint8_t apers) {
    _write_reg(REG_PERS, ((ppers & 0x0F) << 4) | (apers & 0x0F));
}

void APDS9960Full::enable_als_interrupt(bool enabled) {
    uint8_t val = _read_reg(REG_ENABLE);
    if (enabled) val |= 0x10;
    else         val &= ~0x10;
    _write_reg(REG_ENABLE, val);
}

void APDS9960Full::enable_proximity_interrupt(bool enabled) {
    uint8_t val = _read_reg(REG_ENABLE);
    if (enabled) val |= 0x20;
    else         val &= ~0x20;
    _write_reg(REG_ENABLE, val);
}

void APDS9960Full::clear_proximity_interrupt() {
    uint8_t reg = REG_PICLEAR;
    _transport.write(&reg, 1);
}

void APDS9960Full::clear_als_interrupt() {
    uint8_t reg = REG_CICLEAR;
    _transport.write(&reg, 1);
}

void APDS9960Full::clear_all_interrupts() {
    uint8_t reg = REG_AICLEAR;
    _transport.write(&reg, 1);
}

void APDS9960Full::set_proximity_offset(int8_t ur, int8_t dl) {
    auto encode = [](int8_t v) -> uint8_t {
        if (v < 0) return 0x80 | ((-v) & 0x7F);
        return v & 0x7F;
    };
    _write_reg(REG_POFFSET_UR, encode(ur));
    _write_reg(REG_POFFSET_DL, encode(dl));
}

void APDS9960Full::set_proximity_mask(bool u, bool d, bool l, bool r) {
    uint8_t c3 = _read_reg(REG_CONFIG3) & 0xF0;
    if (u) c3 |= 0x08;
    if (d) c3 |= 0x04;
    if (l) c3 |= 0x02;
    if (r) c3 |= 0x01;
    _write_reg(REG_CONFIG3, c3);
}

void APDS9960Full::enable_gesture(bool enabled) {
    uint8_t val = _read_reg(REG_ENABLE);
    if (enabled) {
        val |= 0x40;
        _write_reg(REG_ENABLE, val);
        uint8_t g4 = _read_reg(REG_GCONF4);
        g4 |= 0x01;
        _write_reg(REG_GCONF4, g4);
    } else {
        val &= ~0x40;
        _write_reg(REG_ENABLE, val);
        uint8_t g4 = _read_reg(REG_GCONF4);
        g4 &= ~0x01;
        _write_reg(REG_GCONF4, g4);
    }
}

void APDS9960Full::configure_gesture(uint8_t ggain, uint8_t gldrive, uint8_t gpulse,
                                      uint8_t gplen, uint8_t gwtime, uint8_t gpenth, uint8_t gexth) {
    _write_reg(REG_GPENTH, gpenth);
    _write_reg(REG_GEXTH, gexth);
    uint8_t g2 = ((ggain & 0x03) << 5) | ((gldrive & 0x03) << 3) | (gwtime & 0x07);
    _write_reg(REG_GCONF2, g2);
    _write_reg(REG_GPULSE, ((gplen & 0x03) << 6) | (gpulse & 0x3F));
}

bool APDS9960Full::gesture_available() {
    return (_read_reg(REG_GSTATUS) & 0x01) != 0;
}

uint8_t APDS9960Full::read_gesture_fifo(uint8_t* buf, uint8_t max_len) {
    uint8_t level = _read_reg(REG_GFLVL);
    if (level == 0 || level > max_len) level = max_len;
    uint8_t reg = REG_GFIFO_U;
    for (uint8_t i = 0; i < level; i++) {
        _transport.write_read(&reg, 1, buf + i * 4, 4);
    }
    return level;
}

uint8_t APDS9960Full::gesture_fifo_level() {
    return _read_reg(REG_GFLVL);
}

void APDS9960Full::clear_gesture_fifo() {
    uint8_t g4 = _read_reg(REG_GCONF4);
    g4 |= 0x04;
    _write_reg(REG_GCONF4, g4);
}

void APDS9960Full::enable_gesture_interrupt(bool enabled) {
    uint8_t g4 = _read_reg(REG_GCONF4);
    if (enabled) g4 |= 0x02;
    else         g4 &= ~0x02;
    _write_reg(REG_GCONF4, g4);
}

uint8_t APDS9960Full::status() {
    return _read_reg(REG_STATUS);
}

bool APDS9960Full::is_als_valid() {
    return (_read_reg(REG_STATUS) & 0x01) != 0;
}

bool APDS9960Full::is_proximity_valid() {
    return (_read_reg(REG_STATUS) & 0x02) != 0;
}

bool APDS9960Full::is_als_saturated() {
    return (_read_reg(REG_STATUS) & 0x80) != 0;
}

bool APDS9960Full::is_proximity_saturated() {
    return (_read_reg(REG_STATUS) & 0x40) != 0;
}

uint8_t APDS9960Full::chip_id() {
    return _read_reg(REG_ID);
}

File: nodejs/packages/periph/src/chips/light/apds9960.js
'use strict';

const _REG_ENABLE       = 0x80;
const _REG_ATIME      = 0x81;
const _REG_WTIME      = 0x83;
const _REG_AILTL      = 0x84;
const _REG_AILTH      = 0x85;
const _REG_AIHTL      = 0x86;
const _REG_AIHTH      = 0x87;
const _REG_PILT       = 0x89;
const _REG_PIHT       = 0x8B;
const _REG_PERS       = 0x8C;
const _REG_CONFIG1    = 0x8D;
const _REG_PPULSE     = 0x8E;
const _REG_CONTROL    = 0x8F;
const _REG_CONFIG2    = 0x90;
const _REG_ID         = 0x92;
const _REG_STATUS     = 0x93;
const _REG_CDATAL     = 0x94;
const _REG_RDATAL     = 0x96;
const _REG_GDATAL     = 0x98;
const _REG_BDATAL     = 0x9A;
const _REG_PDATA      = 0x9C;
const _REG_POFFSET_UR = 0x9D;
const _REG_POFFSET_DL = 0x9E;
const _REG_CONFIG3    = 0x9F;
const _REG_GPENTH     = 0xA0;
const _REG_GEXTH      = 0xA1;
const _REG_GCONF1     = 0xA2;
const _REG_GCONF2     = 0xA3;
const _REG_GOFFSET_U  = 0xA4;
const _REG_GOFFSET_D  = 0xA5;
const _REG_GPULSE     = 0xA6;
const _REG_GOFFSET_L  = 0xA7;
const _REG_GOFFSET_R  = 0xA9;
const _REG_GCONF3     = 0xAA;
const _REG_GCONF4     = 0xAB;
const _REG_GFLVL      = 0xAE;
const _REG_GSTATUS    = 0xAF;
const _REG_PICLEAR    = 0xE5;
const _REG_CICLEAR    = 0xE6;
const _REG_AICLEAR    = 0xE7;
const _REG_GFIFO_U    = 0xFC;

const _ATIME_DEFAULT   = 0xB6;
const _CONTROL_DEFAULT = 0x01;
const _CONFIG2_DEFAULT = 0x01;

function sleep(ms) {
    const end = Date.now() + ms;
    while (Date.now() < end) {}
}

/**
 * APDS-9960 digital proximity, ambient light, RGB and gesture sensor — minimal interface.
 *
 * Provides ambient light and color (RGBC) readings with no configuration
 * beyond the transport. The ALS/Color engine is enabled at construction
 * with sensible defaults.
 *
 * Default configuration (written at construction):
 * - ATIME = 0xB6 (72 cycles, ~200 ms integration, max count 65535)
 * - AGAIN = 1 (4x ALS gain)
 * - CONFIG2 = 0x01 (LED_BOOST=100%, reserved bit 0 set)
 * - PON + AEN enabled; no wait, proximity, gesture, or interrupts
 */
class APDS9960Minimal {
    /**
     * @param {object} transport - Configured I2C transport pointing at the device (address 0x39).
     */
    constructor(transport) {
        this._transport = transport;
        sleep(6);
        const id = this._readReg(_REG_ID);
        if (id !== 0xAB) throw new Error('APDS-9960 not found (ID=0x' + id.toString(16) + ', expected 0xAB)');
        this._writeReg(_REG_ENABLE, 0x00);
        this._writeReg(_REG_ATIME, _ATIME_DEFAULT);
        this._writeReg(_REG_CONTROL, _CONTROL_DEFAULT);
        this._writeReg(_REG_CONFIG2, _CONFIG2_DEFAULT);
        this._writeReg(_REG_ENABLE, 0x03);
        sleep(210);
    }

    _writeReg(reg, value) {
        this._transport.write(Buffer.from([reg, value]));
    }

    _readReg(reg) {
        return this._transport.writeRead(Buffer.from([reg]), 1)[0];
    }

    _readReg16LE(reg) {
        const buf = this._transport.writeRead(Buffer.from([reg]), 2);
        return buf[0] | (buf[1] << 8);
    }

    /**
     * Read the clear (unfiltered) channel.
     * @returns {number} Raw clear channel count, 0-65535.
     */
    colorClear() { return this._readReg16LE(_REG_CDATAL); }

    /**
     * Read the red channel.
     *
     * Burst-reads all 8 bytes from CDATAL to trigger the atomic latch.
     * @returns {number} Raw red channel count, 0-65535.
     */
    colorRed() {
        const raw = this._transport.writeRead(Buffer.from([_REG_CDATAL]), 8);
        return raw[2] | (raw[3] << 8);
    }

    /**
     * Read the green channel.
     *
     * Burst-reads all 8 bytes from CDATAL to trigger the atomic latch.
     * @returns {number} Raw green channel count, 0-65535.
     */
    colorGreen() {
        const raw = this._transport.writeRead(Buffer.from([_REG_CDATAL]), 8);
        return raw[4] | (raw[5] << 8);
    }

    /**
     * Read the blue channel.
     *
     * Burst-reads all 8 bytes from CDATAL to trigger the atomic latch.
     * @returns {number} Raw blue channel count, 0-65535.
     */
    colorBlue() {
        const raw = this._transport.writeRead(Buffer.from([_REG_CDATAL]), 8);
        return raw[6] | (raw[7] << 8);
    }

    /**
     * Read all four RGBC channels in one burst.
     *
     * Reading CDATAL at 0x94 atomically latches all eight bytes 0x94-0x9B.
     *
     * @returns {{ clear: number, red: number, green: number, blue: number }}
     */
    color() {
        const raw = this._transport.writeRead(Buffer.from([_REG_CDATAL]), 8);
        return {
            clear: raw[0] | (raw[1] << 8),
            red:   raw[2] | (raw[3] << 8),
            green: raw[4] | (raw[5] << 8),
            blue:  raw[6] | (raw[7] << 8)
        };
    }
}

/**
 * APDS-9960 full interface — extends APDS9960Minimal with proximity, gesture, and configuration.
 *
 * Adds proximity detection, gesture engine, wait engine, threshold and
 * interrupt configuration, status queries, and device identification.
 */
class APDS9960Full extends APDS9960Minimal {
    /**
     * @param {object} transport - Configured I2C transport pointing at the device (address 0x39).
     */
    constructor(transport) {
        super(transport);
    }

    /**
     * Enable or disable the proximity engine.
     * @param {boolean} enabled - true to enable PEN, false to disable.
     */
    enableProximity(enabled) {
        let val = this._readReg(_REG_ENABLE);
        if (enabled) val |= 0x04; else val &= ~0x04;
        this._writeReg(_REG_ENABLE, val);
    }

    /**
     * Read the proximity count.
     * @returns {number} Proximity count 0-255; higher means closer.
     */
    proximity() { return this._readReg(_REG_PDATA); }

    /**
     * Enable or disable the wait engine.
     * @param {boolean} enabled - true to enable WEN, false to disable.
     */
    enableWait(enabled) {
        let val = this._readReg(_REG_ENABLE);
        if (enabled) val |= 0x08; else val &= ~0x08;
        this._writeReg(_REG_ENABLE, val);
    }

    /**
     * Configure the wait time between ALS/proximity cycles.
     * @param {number} wtime - WTIME register value 0-255.
     * @param {boolean} [wlong=false] - true to enable WLONG 12x multiplier.
     */
    configureWait(wtime, wlong = false) {
        this._writeReg(_REG_WTIME, wtime & 0xFF);
        let c1 = this._readReg(_REG_CONFIG1);
        if (wlong) c1 |= 0x02; else c1 &= ~0x02;
        c1 = (c1 & 0x03) | 0x60;
        this._writeReg(_REG_CONFIG1, c1);
    }

    /**
     * Configure ALS integration time and gain.
     * @param {number} atime - ATIME register value 0-255.
     * @param {number} again - ALS gain 0-3 (0=1x, 1=4x, 2=16x, 3=64x).
     */
    configureAls(atime, again) {
        this._writeReg(_REG_ATIME, atime & 0xFF);
        let ctrl = this._readReg(_REG_CONTROL);
        ctrl = (ctrl & 0xFC) | (again & 0x03);
        this._writeReg(_REG_CONTROL, ctrl);
    }

    /**
     * Configure proximity LED drive, gain, pulse count and length.
     * @param {number} ldrive - LED drive strength 0-3.
     * @param {number} pgain - Proximity gain 0-3.
     * @param {number} ppulse - Pulse count minus 1, 0-63.
     * @param {number} pplen - Pulse length 0-3.
     */
    configureProximityLed(ldrive, pgain, ppulse, pplen) {
        let ctrl = this._readReg(_REG_CONTROL);
        ctrl = ((ldrive & 0x03) << 6) | ((pgain & 0x03) << 2) | (ctrl & 0x03);
        this._writeReg(_REG_CONTROL, ctrl);
        this._writeReg(_REG_PPULSE, ((pplen & 0x03) << 6) | (ppulse & 0x3F));
    }

    /**
     * Set additional LED current boost.
     * @param {number} boost - LED_BOOST 0-3 (0=100%, 1=150%, 2=200%, 3=300%).
     */
    setLedBoost(boost) {
        let c2 = this._readReg(_REG_CONFIG2);
        c2 = (c2 & 0xCF) | ((boost & 0x03) << 4) | 0x01;
        this._writeReg(_REG_CONFIG2, c2);
    }

    /**
     * Set ALS interrupt thresholds.
     * @param {number} low - Low threshold 0-65535.
     * @param {number} high - High threshold 0-65535.
     */
    alsThreshold(low, high) {
        this._writeReg(_REG_AILTL, low & 0xFF);
        this._writeReg(_REG_AILTH, (low >> 8) & 0xFF);
        this._writeReg(_REG_AIHTL, high & 0xFF);
        this._writeReg(_REG_AIHTH, (high >> 8) & 0xFF);
    }

    /**
     * Set proximity interrupt thresholds.
     * @param {number} low - Low threshold 0-255.
     * @param {number} high - High threshold 0-255.
     */
    proximityThreshold(low, high) {
        this._writeReg(_REG_PILT, low & 0xFF);
        this._writeReg(_REG_PIHT, high & 0xFF);
    }

    /**
     * Set interrupt persistence filters.
     * @param {number} ppers - Proximity persistence 0-15.
     * @param {number} apers - ALS persistence 0-15.
     */
    setPersistence(ppers, apers) {
        this._writeReg(_REG_PERS, ((ppers & 0x0F) << 4) | (apers & 0x0F));
    }

    /**
     * Enable or disable ALS interrupt.
     * @param {boolean} enabled - true to enable AIEN, false to disable.
     */
    enableAlsInterrupt(enabled) {
        let val = this._readReg(_REG_ENABLE);
        if (enabled) val |= 0x10; else val &= ~0x10;
        this._writeReg(_REG_ENABLE, val);
    }

    /**
     * Enable or disable proximity interrupt.
     * @param {boolean} enabled - true to enable PIEN, false to disable.
     */
    enableProximityInterrupt(enabled) {
        let val = this._readReg(_REG_ENABLE);
        if (enabled) val |= 0x20; else val &= ~0x20;
        this._writeReg(_REG_ENABLE, val);
    }

    /**
     * Clear the proximity interrupt via address-only write to PICLEAR.
     */
    clearProximityInterrupt() {
        this._transport.write(Buffer.from([_REG_PICLEAR]));
    }

    /**
     * Clear the ALS/color interrupt via address-only write to CICLEAR.
     */
    clearAlsInterrupt() {
        this._transport.write(Buffer.from([_REG_CICLEAR]));
    }

    /**
     * Clear all non-gesture interrupts via address-only write to AICLEAR.
     */
    clearAllInterrupts() {
        this._transport.write(Buffer.from([_REG_AICLEAR]));
    }

    /**
     * Set proximity offset for UP/RIGHT and DOWN/LEFT photodiodes.
     * @param {number} ur - UP/RIGHT offset -127 to +127 (sign-magnitude).
     * @param {number} dl - DOWN/LEFT offset -127 to +127 (sign-magnitude).
     */
    setProximityOffset(ur, dl) {
        const encode = (v) => v < 0 ? (0x80 | ((-v) & 0x7F)) : (v & 0x7F);
        this._writeReg(_REG_POFFSET_UR, encode(ur));
        this._writeReg(_REG_POFFSET_DL, encode(dl));
    }

    /**
     * Mask individual photodiodes in proximity detection.
     * @param {boolean} u - true to mask UP.
     * @param {boolean} d - true to mask DOWN.
     * @param {boolean} l - true to mask LEFT.
     * @param {boolean} r - true to mask RIGHT.
     */
    setProximityMask(u, d, l, r) {
        let c3 = this._readReg(_REG_CONFIG3) & 0xF0;
        if (u) c3 |= 0x08;
        if (d) c3 |= 0x04;
        if (l) c3 |= 0x02;
        if (r) c3 |= 0x01;
        this._writeReg(_REG_CONFIG3, c3);
    }

    /**
     * Enable or disable the gesture engine.
     * @param {boolean} enabled - true to enable GEN and set GMODE, false to disable.
     */
    enableGesture(enabled) {
        let val = this._readReg(_REG_ENABLE);
        if (enabled) {
            val |= 0x40;
            this._writeReg(_REG_ENABLE, val);
            let g4 = this._readReg(_REG_GCONF4);
            g4 |= 0x01;
            this._writeReg(_REG_GCONF4, g4);
        } else {
            val &= ~0x40;
            this._writeReg(_REG_ENABLE, val);
            let g4 = this._readReg(_REG_GCONF4);
            g4 &= ~0x01;
            this._writeReg(_REG_GCONF4, g4);
        }
    }

    /**
     * Configure gesture engine parameters.
     * @param {number} ggain - Gesture gain 0-3.
     * @param {number} gldrive - Gesture LED drive 0-3.
     * @param {number} gpulse - Gesture pulse count minus 1, 0-63.
     * @param {number} gplen - Gesture pulse length 0-3.
     * @param {number} gwtime - Gesture wait time 0-7.
     * @param {number} gpenth - Gesture proximity entry threshold 0-255.
     * @param {number} gexth - Gesture exit threshold 0-255.
     */
    configureGesture(ggain, gldrive, gpulse, gplen, gwtime, gpenth, gexth) {
        this._writeReg(_REG_GPENTH, gpenth & 0xFF);
        this._writeReg(_REG_GEXTH, gexth & 0xFF);
        const g2 = ((ggain & 0x03) << 5) | ((gldrive & 0x03) << 3) | (gwtime & 0x07);
        this._writeReg(_REG_GCONF2, g2);
        this._writeReg(_REG_GPULSE, ((gplen & 0x03) << 6) | (gpulse & 0x3F));
    }

    /**
     * Check if gesture data is available in the FIFO.
     * @returns {boolean} true if GSTATUS.GVALID is set.
     */
    gestureAvailable() { return !!(this._readReg(_REG_GSTATUS) & 0x01); }

    /**
     * Read all gesture datasets from the FIFO.
     * @returns {Array<{ u: number, d: number, l: number, r: number }>}
     */
    readGestureFifo() {
        const level = this._readReg(_REG_GFLVL);
        if (level === 0) return [];
        const result = [];
        for (let i = 0; i < level; i++) {
            const raw = this._transport.writeRead(Buffer.from([_REG_GFIFO_U]), 4);
            result.push({ u: raw[0], d: raw[1], l: raw[2], r: raw[3] });
        }
        return result;
    }

    /**
     * Read the number of datasets in the gesture FIFO.
     * @returns {number} Number of 4-byte datasets currently in FIFO.
     */
    gestureFifoLevel() { return this._readReg(_REG_GFLVL); }

    /**
     * Clear the gesture FIFO by setting GFIFO_CLR in GCONF4.
     */
    clearGestureFifo() {
        let g4 = this._readReg(_REG_GCONF4);
        g4 |= 0x04;
        this._writeReg(_REG_GCONF4, g4);
    }

    /**
     * Enable or disable gesture interrupt.
     * @param {boolean} enabled - true to enable GIEN, false to disable.
     */
    enableGestureInterrupt(enabled) {
        let g4 = this._readReg(_REG_GCONF4);
        if (enabled) g4 |= 0x02; else g4 &= ~0x02;
        this._writeReg(_REG_GCONF4, g4);
    }

    /**
     * Read the raw STATUS register.
     * @returns {number} Raw STATUS byte.
     */
    status() { return this._readReg(_REG_STATUS); }

    /**
     * Check if ALS/color data is valid.
     * @returns {boolean} true if STATUS.AVALID is set.
     */
    isAlsValid() { return !!(this._readReg(_REG_STATUS) & 0x01); }

    /**
     * Check if proximity data is valid.
     * @returns {boolean} true if STATUS.PVALID is set.
     */
    isProximityValid() { return !!(this._readReg(_REG_STATUS) & 0x02); }

    /**
     * Check if the clear photodiode is saturated.
     * @returns {boolean} true if STATUS.CPSAT is set.
     */
    isAlsSaturated() { return !!(this._readReg(_REG_STATUS) & 0x80); }

    /**
     * Check if analog saturation occurred during proximity.
     * @returns {boolean} true if STATUS.PGSAT is set.
     */
    isProximitySaturated() { return !!(this._readReg(_REG_STATUS) & 0x40); }

    /**
     * Read the device ID register.
     * @returns {number} ID register value (expect 0xAB).
     */
    chipId() { return this._readReg(_REG_ID); }
}

module.exports = { APDS9960Minimal, APDS9960Full };
